### PR TITLE
Item icons

### PR DESCRIPTION
**Pull request by Zsteven44**

### CODE REVIEW

File: app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package tech.zafrani.pubgapp.adapters.decorations;
+
+import android.graphics.Canvas;
+import android.graphics.Rect;
+import android.graphics.drawable.Drawable;
+import android.support.v7.widget.RecyclerView;
+import android.view.View;
+
+public class ItemRowDecoration extends RecyclerView.ItemDecoration{
+    private Drawable divider;
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: final

Comment (zafrani) [reply app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: @NonNull

File: app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    public ItemRowDecoration(Drawable divider) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: @Nonnull final

File: app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        this.divider = divider;
+    }
+
+    @Override
+    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: final all these

File: app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+        if (parent.getChildAdapterPosition(view) == 0) {
+            return;
+        }
+
+        outRect.top = divider.getIntrinsicHeight();
+    }
+
+    @Override
+    public void onDraw(Canvas canvas, RecyclerView parent, RecyclerView.State state) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: Final all of these

File: app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        int dividerRight = parent.getWidth() - parent.getPaddingRight();
+
+        int childCount = parent.getChildCount();
+        for (int i = 0; i < childCount - 1; i++) {
+            View child = parent.getChildAt(i);
+
+            RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
+
+            int dividerTop = child.getBottom() + params.bottomMargin;
+            int dividerBottom = dividerTop + divider.getIntrinsicHeight();
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: final these and the above variables

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 6]
```diff
@@ -1,35 +1,127 @@
...
     private TextView categoryTextView;
+    private TextView ammoTextView;
+    private TextView magazineTextView;
+    private TextView capacityTextView;
+    private TextView damageTextView;
+    private TextView rangeTextView;
+    private TextView stabilityTextView;
+    private TextView rateTextView;
+    private ImageView imageImageView;
+    private Context context;
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: final all of these variables

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 7]
```diff
@@ -1,35 +1,127 @@
...
+
     public void setItemCategory(@NonNull final String itemCategory) {
         this.categoryTextView.setText(itemCategory);
     }
+
+    public void setItemAmmo(@Nullable final String itemAmmo) {
+        if (itemAmmo != null) {
+            this.ammoTextView.setText(context.getString(R.string.row_item_ammo, itemAmmo));
+        } else {
+            this.ammoTextView.setText(" - ");
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: Make this value a static variable of the class

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 8]
```diff
@@ -1,35 +1,127 @@
...
+        }
+    }
+    public void setItemMagazine(final Float itemMagazine) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: @Nullable

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 9]
```diff
@@ -1,35 +1,127 @@
...
+        if (itemMagazine != null) {
+            this.magazineTextView.setText(context.getString(R.string.row_item_magazine, itemMagazine));
+        }else {
+            this.magazineTextView.setText(" - ");
+        }
+    }
+    public void setItemCapacity(final Float itemCapacity) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: @Nullable

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 10]
```diff
@@ -1,35 +1,127 @@
...
+            if(itemDamage.containsKey("chest0")){
+                this.damageTextView.setText(context.getString(R.string.row_item_damage, itemDamage.get("chest0")));
+            } else {
+                this.damageTextView.setText(context.getString(R.string.row_item_damage, itemDamage.get("base")));
+            }
+        }else {
+            this.damageTextView.setText(" - ");
+        }
+    }
+    public void setItemStability(final Float itemStability) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: @Nullable

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 11]
```diff
@@ -1,35 +1,127 @@
...
+        if (itemStability!= null) {
+            this.stabilityTextView.setText(context.getString(R.string.row_item_stability, itemStability));
+        }else {
+            this.stabilityTextView.setText(" - ");
+        }
+    }
+    public void setItemRange(final Float itemRange) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: @Nullable

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 12]
```diff
@@ -1,35 +1,127 @@
...
+        if (itemRange!= null) {
+            this.rangeTextView.setText(context.getString(R.string.row_item_range, itemRange));
+        }else {
+            this.rangeTextView.setText(" - ");
+        }
+    }
+    public void setItemRate(final Float itemRate) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: @Nullable

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 13]
```diff
@@ -1,35 +1,127 @@
...
+        if (itemRate != null) {
+            this.rateTextView.setText(context.getString(R.string.row_item_rate, itemRate));
+        }else {
+            this.rateTextView.setText(" - ");
+        }
+    }
+    public void setItemIcon(final String itemImage, String category) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: @Nullable

File: app/src/main/java/tech/zafrani/pubgapp/fragments/ItemTabFragment.java [hunk 14]
```diff
@@ -60,12 +62,15 @@ public void onViewCreated(@NonNull final View view,
...
         for (final Type type : category.getTypes()) {
             Log.e(getClass()
-                    .getSimpleName(), "Items for Type " + type + ": " + (type.getItems() == null ? "null" : type.getItems().toString()));
+                    .getSimpleName(), "Items for Type " + type + ": " +  type.getItems().toString());
             items.addAll(type.getItems());
         }
         recyclerView = (RecyclerView) view.findViewById(R.id.fragment_itemtab_recyclerview);
         recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
-        final ItemListAdapter adapter = new ItemListAdapter(items);
+        Drawable dividerDrawable = getResources().getDrawable(R.drawable.item_row_divider_decoration);
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/fragments/ItemTabFragment.java]: final

File: app/src/main/java/tech/zafrani/pubgapp/fragments/ItemTabFragment.java [hunk 15]
```diff
@@ -60,12 +62,15 @@ public void onViewCreated(@NonNull final View view,
...
+        RecyclerView.ItemDecoration itemDividerDecoration = new ItemRowDecoration(dividerDrawable);
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/fragments/ItemTabFragment.java]: final

File: app/src/main/java/tech/zafrani/pubgapp/utils/FileUtil.java [hunk 16]
```diff
@@ -26,12 +28,20 @@ public static Items getItems(@NonNull final Activity activity) throws IOExceptio
...
+            return bindItems(gson.fromJson(reader, Items.class));
 
         }
         return null;
     }
 
+    public static Items bindItems(@NonNull Items items) {
+        List<Category> categories = items.getCategories();
+        for (Category category: categories){
+            category.updateChildren();
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/utils/FileUtil.java]: why is this needed?

File: app/src/main/res/drawable/item_row_divider_decoration.xml [hunk 17]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <size android:height="1dp" />
+    <solid android:color="#aaaaaa" />
```
Comment (zafrani) [inline app/src/main/res/drawable/item_row_divider_decoration.xml]: Add this to colors.xml

**Review by zafrani — CHANGES_REQUESTED** [review]

File: app/src/main/java/tech/zafrani/pubgapp/utils/FileUtil.java (hunk 16 above)
Comment (Zsteven44) [reply app/src/main/java/tech/zafrani/pubgapp/utils/FileUtil.java]: The constructor wasn't running updateChildren.

Comment (zafrani) [reply app/src/main/java/tech/zafrani/pubgapp/utils/FileUtil.java]: Make the parameter, list and each item final.

File: app/src/main/java/tech/zafrani/pubgapp/adapters/ItemListAdapter.java [hunk 18]
```diff
@@ -35,6 +35,14 @@ public void onBindViewHolder(@NonNull final ItemViewHolder holder, @NonNull fina
...
         holder.setItemType(item.getType());
         holder.setItemCategory(item.getCategory());
+        holder.setItemAmmo(item.getAmmo());
+        holder.setItemMagazine(item.getMagazine());
+        holder.setItemCapacity(item.getCapacity());
+        holder.setItemDamage(item.getDamage());
+        holder.setItemRange(item.getRange());
+        holder.setItemRate(item.getRate());
+        holder.setItemStability(item.getStability());
+        holder.setItemIcon(item.getImageUrl(), item.getCategory());
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/ItemListAdapter.java:45]: why not use a holder.setItem function and set all of these from it?

File: app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    public void onDraw(final Canvas canvas,
+                       final RecyclerView parent,
+                       final RecyclerView.State state) {
+
+        final int dividerLeft = parent.getPaddingLeft();
+        final int dividerRight = parent.getWidth() - parent.getPaddingRight();
+
+        final int childCount = parent.getChildCount();
+        for (int i = 0; i < childCount - 1; i++) {
+            View child = parent.getChildAt(i);
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: final

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 20]
```diff
@@ -1,35 +1,128 @@
...
+
 import tech.zafrani.pubgapp.R;
 
 
 public class ItemViewHolder  extends RecyclerView.ViewHolder{
-    private TextView nameTextView;
-    private TextView typeTextView;
-    private TextView categoryTextView;
+    private final String IMAGE_URL_BASE= "https://github.com/Zsteven44/PUBG-Companion-Images/blob/master/images/";
+    private final String ARG_RAW = "?raw=true";
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: this can be static

File: app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java [hunk 21]
```diff
@@ -1,35 +1,128 @@
...
+import java.util.HashMap;
+
 import tech.zafrani.pubgapp.R;
 
 
 public class ItemViewHolder  extends RecyclerView.ViewHolder{
-    private TextView nameTextView;
-    private TextView typeTextView;
-    private TextView categoryTextView;
+    private final String IMAGE_URL_BASE= "https://github.com/Zsteven44/PUBG-Companion-Images/blob/master/images/";
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/adapters/viewholders/ItemViewHolder.java]: this can be static

File: app/src/main/java/tech/zafrani/pubgapp/models/Item.java [hunk 22]
```diff
@@ -17,13 +20,49 @@
...
+    private final Float magazine;
+    @SerializedName("capacity")
+    private final Float capacity;
+    @Nullable
+    @SerializedName("damage")
+    private final HashMap<String, String> damage;
+    @SerializedName("stability")
+    private final Float stability;
+    @SerializedName("rate")
+    private final Float rate;
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/models/Item.java]: These should be nullable.

File: app/src/main/java/tech/zafrani/pubgapp/utils/FileUtil.java [hunk 23]
```diff
@@ -26,12 +28,20 @@ public static Items getItems(@NonNull final Activity activity) throws IOExceptio
...
             final BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
 
-            return gson.fromJson(reader, Items.class);
+            return bindItems(gson.fromJson(reader, Items.class));
 
         }
         return null;
     }
 
+    public static Items bindItems(@NonNull final Items items) {
```
Comment (zafrani) [inline app/src/main/java/tech/zafrani/pubgapp/utils/FileUtil.java]: This should be non null

File: app/src/main/res/drawable/item_row_divider_decoration.xml [hunk 24]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <size android:height="1dp" />
+    <solid android:color="@color/colorDivider" />
```
Comment (zafrani) [inline app/src/main/res/drawable/item_row_divider_decoration.xml:5]: This is in the color file so it doesn't need to have color in its value. It can just be divider.

File: app/src/main/res/drawable/item_view_selector.xml [hunk 25]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:background="@color/colorPrimary" android:state_selected="true">
+
+    </item>
+    <item android:background="@color/colorPrimaryDark" android:state_pressed="true">
```
Comment (zafrani) [inline app/src/main/res/drawable/item_view_selector.xml]: Same here, this doesn't need color in the value.

**Review by zafrani — CHANGES_REQUESTED** [review]

File: app/src/main/java/tech/zafrani/pubgapp/adapters/ItemListAdapter.java (hunk 18 above)
Comment (Zsteven44) [reply app/src/main/java/tech/zafrani/pubgapp/adapters/ItemListAdapter.java:45]: this changes in the next branch I pull request to a bind() function

File: app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java (hunk 19 above)
Comment (Zsteven44) [reply app/src/main/java/tech/zafrani/pubgapp/adapters/decorations/ItemRowDecoration.java]: removed, replaced with view in xml.

File: app/src/main/res/drawable/item_row_divider_decoration.xml (hunk 24 above)
Comment (Zsteven44) [reply app/src/main/res/drawable/item_row_divider_decoration.xml:5]: removing this

File: app/src/main/res/drawable/item_view_selector.xml (hunk 25 above)
Comment (Zsteven44) [reply app/src/main/res/drawable/item_view_selector.xml]: removed becuase its no longer being used.

**Review by zafrani — APPROVED** [review]
